packages/chain-bridge/src/router/BridgeRouter.sol

```diff
@@ -0,0 +1,828 @@
+        (nativeFee, tokenFee) = IBridgeAdapter(specifiedAdapter).estimateFee(
+            destinationChainId,
+            asset,
+            amount,
+            options.adapterParams,
+            operationType
+        );
+
+        return (nativeFee, tokenFee, specifiedAdapter);
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function quote(
+        uint16 destinationChainId,
+        address asset,
+        uint256 amount,
+        BridgeTypes.BridgeOptions calldata options,
+        BridgeTypes.OperationType operationType
+    )
+        external
+        view
+        returns (uint256 nativeFee, uint256 tokenFee, address specifiedAdapter)
+    {
+        // Get the base fee from internal quote
+        (uint256 baseFee, uint256 baseTokenFee, address adapter) = _quote(
+            destinationChainId,
+            asset,
+            amount,
+            options,
+            operationType
+        );
+
+        // Apply fee buffer to account for fee volatility
+        uint256 bufferedNativeFee = _applyFeeBuffer(baseFee);
+
+        return (bufferedNativeFee, baseTokenFee, adapter);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                        ADAPTER CALLBACK FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IBridgeRouter
+    function updateOperationStatus(
+        bytes32 operationId,
+        BridgeTypes.OperationStatus status
+    ) external onlyRegisteredAdapter {
+        if (operationToAdapter[operationId] != msg.sender)
+            revert Unauthorized();
+
+        // Allow transitions from QUEUED to SENT, or from SENT to FAILED
+        if (
+            status == BridgeTypes.OperationStatus.SENT &&
+            operationStatuses[operationId] == BridgeTypes.OperationStatus.QUEUED
+        ) {
+            operationStatuses[operationId] = status;
+            emit OperationStatusUpdated(operationId, status);
+        } else if (
+            status == BridgeTypes.OperationStatus.FAILED &&
+            operationStatuses[operationId] == BridgeTypes.OperationStatus.SENT
+        ) {
+            operationStatuses[operationId] = status;
+            emit OperationStatusUpdated(operationId, status);
+        }
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function updateReceiveStatus(
+        bytes32 requestId,
+        address recipient,
+        BridgeTypes.OperationStatus status
+    ) external onlyRegisteredAdapter {
+        requestReceivedByAdapter[requestId] = msg.sender;
+
+        // Only update status if it's a failure
+        if (status == BridgeTypes.OperationStatus.FAILED) {
+            operationStatuses[requestId] = status;
+            emit OperationStatusUpdated(requestId, status);
+        }
+
+        // Always emit delivery event
+        emit MessageDelivered(
+            requestId,
+            recipient,
+            status != BridgeTypes.OperationStatus.FAILED
+        );
```

Comment:
the name of the event is confusing, as it suggests the message was delivered while it cointains a flag `delivered`